globalping/auth.go

```diff
@@ -0,0 +1,268 @@
+package globalping
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+type Token struct {
+	// AccessToken is the token that authorizes and authenticates
+	// the requests.
+	AccessToken string `json:"access_token"`
+
+	// TokenType is the type of token.
+	// The Type method returns either this or "Bearer", the default.
+	TokenType string `json:"token_type,omitempty"`
+
+	// RefreshToken is a token that's used by the application
+	// (as opposed to the user) to refresh the access token
+	// if it expires.
+	RefreshToken string `json:"refresh_token,omitempty"`
+
+	// Expiry is the optional expiration time of the access token.
+	//
+	// If zero, TokenSource implementations will reuse the same
+	// token forever and RefreshToken or equivalent
+	// mechanisms for that TokenSource will not be used.
+	Expiry time.Time `json:"expiry,omitempty"`
+}
+
+type AuthorizeError struct {
+	Code        int
+	ErrorCode   string
+	Description string
+}
+
+func (e *AuthorizeError) Error() string {
+	return fmt.Sprintf("%s: %s", e.ErrorCode, e.Description)
+}
+
+func (c *client) Authorize(callback func(error)) string {
+	pkce := oauth2.GenerateVerifier()
+	mux := http.NewServeMux()
+	server := &http.Server{
+		Handler: mux,
+	}
+	mux.HandleFunc("/callback", func(w http.ResponseWriter, req *http.Request) {
+		req.ParseForm()
+		token, err := c.exchange(req.Form, pkce, "http://"+server.Addr+"/callback")
+		if err != nil {
+			http.Redirect(w, req, c.dashboardURL+"/authorize/error", http.StatusFound)
+		} else {
+			http.Redirect(w, req, c.dashboardURL+"/authorize/success", http.StatusFound)
+		}
+		go func() {
+			server.Shutdown(req.Context())
+			if err == nil {
+				c.token.Store(token)
+				if c.onTokenRefresh != nil {
+					c.onTokenRefresh(mapToken(token))
+				}
+			}
+			callback(err)
+		}()
+	})
+	go func() {
+		ports := []int{60000, 60010, 60020, 60030, 60040, 60100, 60110, 60120, 60130, 60140}
+		var err error
+		for _, port := range ports {
+			server.Addr = fmt.Sprintf("localhost:%d", port)
+			err = server.ListenAndServe()
+			if err == nil {
+				break
+			}
+		}
```

Comment:
Not sure if it's an issue with this or with tests, but on Windows (also on Linux via WSL) the tests are currently failing for me with:

```
?       github.com/jsdelivr/globalping-cli      [no test files]
?       github.com/jsdelivr/globalping-cli/globalping/probe     [no test files]
?       github.com/jsdelivr/globalping-cli/mocks        [no test files]
?       github.com/jsdelivr/globalping-cli/packagecloud [no test files]
?       github.com/jsdelivr/globalping-cli/utils        [no test files]
?       github.com/jsdelivr/globalping-cli/version      [no test files]
ok      github.com/jsdelivr/globalping-cli/cmd  (cached)
--- FAIL: Test_Authorize (0.00s)
    auth_test.go:78: Post "http://localhost:60000/callback?code=cod3": dial tcp 127.0.0.1:60000: connect: connection refused
FAIL
FAIL    github.com/jsdelivr/globalping-cli/globalping   0.277s
ok      github.com/jsdelivr/globalping-cli/storage      (cached)
ok      github.com/jsdelivr/globalping-cli/view (cached)
FAIL
```

Reply:
It should work now